components/common-go/grpc/ratelimit.go

```diff
@@ -6,60 +6,143 @@ package grpc
-			Block: fnc.Block,
-			L:     rate.NewLimiter(rate.Every(time.Duration(fnc.RefillInterval)), int(fnc.BucketSize)),
+		var (
+			keyedLimit *lru.Cache
+			key        keyFunc
+		)
+		if fnc.Key != "" && fnc.KeyCacheSize > 0 {
+			keyedLimit, _ = lru.New(int(fnc.KeyCacheSize))
+			key = fieldAccessKey(fnc.Key)
+		}
+
+		funcs[name] = &ratelimitedFunction{
+			RateLimit:   fnc,
+			GlobalLimit: fnc.Limiter(),
+			Key:         key,
+			KeyedLimit:  keyedLimit,
 		}
 	}
 	return funcs
 }
 
+func fieldAccessKey(key string) keyFunc {
+	return func(req interface{}) (string, error) {
+		msg, ok := req.(proto.Message)
+		if !ok {
+			return "", status.Errorf(codes.Internal, "request was not a protobuf message")
+		}
+
+		val, ok := getFieldValue(msg.ProtoReflect(), strings.Split(key, "."))
+		if !ok {
+			return "", status.Errorf(codes.Internal, "Field %s does not exist in message. This is a rate limiting configuration error.", key)
+		}
+		return val, nil
+	}
+}
+
+func getFieldValue(msg protoreflect.Message, path []string) (val string, ok bool) {
+	if len(path) == 0 {
+		return "", false
+	}
+
+	field := msg.Descriptor().Fields().ByName(protoreflect.Name(path[0]))
+	if field == nil {
+		return "", false
+	}
+	if len(path) > 1 {
+		if field.Kind() != protoreflect.MessageKind {
+			// we should go deeper but the field is not a message
+			return "", false
+		}
+		child := msg.Get(field).Message()
+		return getFieldValue(child, path[1:])
+	}
+
+	if field.Kind() != protoreflect.StringKind {
+		// we only support string fields
+		return "", false
+	}
+
+	return msg.Get(field).String(), true
+}
+
 // RatelimitingInterceptor limits how often a gRPC function may be called. If the limit has been
 // exceeded, we'll return resource exhausted.
-type RatelimitingInterceptor map[string]ratelimitedFunction
+type RatelimitingInterceptor map[string]*ratelimitedFunction
 
 type ratelimitedFunction struct {
-	Block bool
-	L     *rate.Limiter
+	RateLimit RateLimit
+
+	GlobalLimit *rate.Limiter
+	Key         keyFunc
+	KeyedLimit  *lru.Cache
 }
 
 // UnaryInterceptor creates a unary interceptor that implements the rate limiting
 func (r RatelimitingInterceptor) UnaryInterceptor() grpc.UnaryServerInterceptor {
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
 		f, ok := r[info.FullMethod]
-		if ok {
-			if f.Block {
-				err := f.L.Wait(ctx)
-				if err == context.Canceled {
-					return nil, err
-				}
-				if err != nil {
-					return nil, status.Error(codes.ResourceExhausted, err.Error())
-				}
-			} else if !f.L.Allow() {
-				return nil, status.Error(codes.ResourceExhausted, "too many requests")
+		if !ok {
+			return handler(ctx, req)
+		}
+
+		var limit *rate.Limiter
+		if f.Key == nil {
+			limit = f.GlobalLimit
+		} else {
+			key, err := f.Key(req)
+			if err != nil {
+				return nil, err
+			}
+
+			f.KeyedLimit.ContainsOrAdd(key, f.RateLimit.Limiter())
```

Comment:
question: this immediately throws creates and trashes a limiter struct on stack AFAIU which could be avoided by not using `ContainsOrAdd`.

Reply:
Agreed - not doing it this way requires extra locking because this operation can happen concurrently. 

We could also ignore the concurrency for the time being and go with a `Peek` followed by an `Add`, running this risk to mis-account for a user during very high call-rate situations.